teps/0011-redirecting-step-inputs-and-outputs.md

```diff
@@ -0,0 +1,143 @@
+
+* Run image `alpine/git` with args `["describe", "--tags"]` in a step and extract the stdout as a Task result without overwriting the image entrypoint to be used by subsequent tasks.
+
+* In general, enable Task authors to access stdout/stderr of a step easily in another step in the same or different Task, without requiring authors to: 1) overwrite image entrypoints; 2) write wrapping scripts solely to do I/O redirection; and 3) build wrapper images to include all needed utilities.
+
+By enabling redirection between steps and Task results, it is much easier for Task authors to composite utility images in a Task.
+
+### Goals
+
+1. Allow the stdout/stderr of a step to be consumed by another step through stdin.
+1. Allow a step to produce a Task result through its stdout or stderr.
+
+### Non-Goals
+
+1. Parse stdout of a step into a structured format (e.g., JSON) and extract information from certain fields.
+
+## Requirements
+
+* Add new fields for a step to specify a file to read from as stdin and files to write to as stdout/stderr.
+* Enhance the `entrypoint` binary to redirect stdin from and stdout/stderr to specified files.
+* Provide examples to use step redirection for the use cases mentioned in [Motivation](#motivation).
+
+## Proposal
+
+The following new fields will be added to the `Step` struct:
+
+```go
+type Step struct {
+    ...
+    // Path to redirect stdin from on container's local filesystem.
+    // +optional
+    Stdin string `json:"stdin,omitempty"`
+    // Path to redirect stdout to on container's local filesystem.
+    // +optional
+    Stdout string `json:"stdout,omitempty"`
+    // Path to redirect stderr to on container's local filesystem.
+    // +optional
+    Stderr string `json:"stderr,omitempty"`
+}
+```
+
+Once `Stdout` or `Stderr` is specified, the corresponding output stream will be duplicated to both the given file and the standard output stream of the container, so users can still view the output through the Pod log API. If both `Stdout` and `Stderr` are set to the same value, the two streams will be written to two file descriptors referring to the same file description, so outputs from both stream will be interleaved in the same file, but there will be no ordering guarantee on the data. The `Stdin` and `Stdout` (or `Stderr`) fields of a step cannot be set to the same value, or the specified file will be truncated and result in an empty stdin.  If multiple steps' `Stdout` fields are set to the same value, the file content will be overwritten by the last outputting step.
+
+Variable substitution will be applied to the new fields, so one could specify `$(results.<name>.path)` to the `Stdout` field to extract the stdout of a step into a Task result. No new variable substitution for accessing the values of `Stdout` and `Stderr` fields will be provided so variable substitution can remain single-pass.
```

Comment:
Given that log output can vary in size dramatically from program to program I'm not sure task results can be promoted as a consistently applicable solution. A user might try to leverage results in this way and get quite frustrated because their container generates so much log data that all of their results are lost along with any other data tekton emits via the termination message. At the moment, blowing through the cap on termination log size will put the TaskRun into a failed state.

Instead I'd like to suggest pursuing a workspaces approach for sharing the data.  I.e. one should specify `$(workspaces.my-logs.path)/stdout` as the target. Other Steps in the Task can access this workspace.  Other Tasks in a Pipeline can be given this workspace (assuming it's a PVC) and also access the stdout data if needed.

You could obviously still use a results path as a desination for the stdout but I think it would have to be done very carefully to avoid blowing up the termination message size, so I'd prefer to push users towards a more broadly-applicable solution like workspaces.

Reply:
I'm going to capture this in the requirements. Thanks!